Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -788,19 +785,78 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
         handledNodes: [name.id]
       )
     }
-    if let unexpected = node.unexpectedBetweenDeinitKeywordAndBody,
-      let signature = unexpected.compactMap({ $0.as(FunctionSignatureSyntax.self) }).only
+    if let unexpected = node.unexpectedBetweenDeinitKeywordAndEffectSpecifiers,
+      let params = unexpected.compactMap({ $0.as(ParameterClauseSyntax.self) }).only
     {
       addDiagnostic(
-        signature,
+        params,
         .deinitCannotHaveParameters,
         fixIts: [
-          FixIt(message: RemoveNodesFixIt(signature), changes: .makeMissing(signature))
+          FixIt(message: RemoveNodesFixIt(params), changes: .makeMissing(params))
+        ],
+        handledNodes: [params.id]
+      )
+    }
+
+    return .visitChildren
+  }
+
+  public override func visit(_ node: DeinitEffectSpecifiersSyntax) -> SyntaxVisitorContinueKind {
+    let throwsTokens: [TokenKind] = [
+      .keyword(.throws),
+      .keyword(.rethrows),
+      .keyword(.try),
+      .keyword(.throw),
+    ]
+    func asThrowingToken(_ syntax: Syntax) -> TokenSyntax? {
+      guard let token = syntax.as(TokenSyntax.self) else { return nil }
+      if token.isMissing { return nil }
+      if throwsTokens.contains(token.tokenKind) { return token }
+      return nil
+    }
+
+    let unexpectedThrows =
+      (node.unexpectedBeforeAsyncSpecifier?.compactMap(asThrowingToken) ?? []) + (node.unexpectedAfterAsyncSpecifier?.compactMap(asThrowingToken) ?? [])
+    if let throwsKeyword = unexpectedThrows.first {
+      addDiagnostic(
+        throwsKeyword,
+        .deinitCannotThrow,
+        fixIts: [
+          FixIt(message: RemoveNodesFixIt(unexpectedThrows), changes: .makeMissing(unexpectedThrows))
         ],
-        handledNodes: [signature.id]
+        handledNodes: unexpectedThrows.map(\.id)
       )
     }
 
+    let isOfSameKind = { AsyncEffectSpecifier(token: $0) != nil }
```

Comment:
`isOfSameKind` is a little misleading here. I would just call it: `isAsyncEffectSpecifier` (and maybe also rename `asThrowingToken` -> `asThrowsEffectSpecifier` for consistency).